src/model.cpp

```diff
@@ -547,4 +547,72 @@ bool Model::fixVariableInterfaces()
     return allOk;
 }
 
+static const size_t MAX_SIZE_T = std::numeric_limits<size_t>::max();
+
+bool isComponentEmpty(const ComponentPtr &component)
+{
+    for (size_t i = component->componentCount() - 1; i != MAX_SIZE_T; --i) {
+        if (isComponentEmpty(component->component(i))) {
+            component->removeComponent(i);
+        }
+    }
+
+    return (component->variableCount() + component->resetCount() + component->componentCount()) == 0
+           && component->math().empty()
+           && !component->isImport();
+}
+
+void checkUnits(const ComponentPtr &component, std::vector<UnitsPtr> &unused)
```

Comment:
This method would also need to check if the units are used in the math (including math in the resets)

Reply:
You're right, bother.  In order to do that, I need to parse the MathML.  If the MathML isn't valid then it can't be parsed, so errors need to be logged.  There is no logger because this is done as part of the `Model` class.  I can move this into the `Printer` class, since that's where people will be using it anyway, and log errors there.  Thoughts?

Reply:
I can almost see this as an optional switch on `printModel` to clean and then print. But might also be easier to simply not do anything with units. Just clean "empty" things for now, cleaning "unused" things can be a future addition if it would ever be useful.

Reply:
In the first round of discussions we talked about that - and concluded somewhere that this should be something separate from the printer in the `Model` class.  I can't remember why!  If I don't check units then it can stay where it is in the `Model`, where it's a little more flexible as you're not constrained to printing (even if that's what most people will do).

Reply:
8837c78c